### PR TITLE
implement Array.prototype.forEach in javascript

### PR DESCRIPTION
**Pull request by yullin-ms**
Micro benchmark result:
before: 169878 ms  
after: 590 ms

Example test code:
var intArr = [];
for (var i = 0; i < 10000; i++) {
    intArr.push(i);
}
for (var i = 0; i < 50000; i++) {
    intArr.forEach(function(element) {});
}

### CODE REVIEW

**Comment by rhuanjl** [comment]
I don’t think that micro-benchmark is right/testing the right thing.

The reason for the massive speed up is that the JIT will  optimise away the callback function completely there so your forEach becomes the toLengtg call and the typeof != function check followed by an empty loop the JIT may even optimise the loop within the forEach away too.

To test the real impact the benchmark needs to use  a callback that does something so can’t be fully optimised away.

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js [hunk 1]
```diff
@@ -238,5 +239,50 @@
...
+        if (__chakraLibrary.isArray(this)) {
+            o = this;
+            len = o.length;
+        } else {
+            o = __chakraLibrary.Object(this);
+            len = __chakraLibrary.GetLength(o);
+        }
+
+        if (typeof callbackfn != "function") {
+            __chakraLibrary.raiseFunctionArgument_NeedFunction("Array.prototype.forEach");
```
Comment (akroshg) [inline lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: could this boilerplate code be refactored with other functions? I assume there will be other implementation in the future which can make use of that.

**Comment by akroshg** [comment]
I agree with you to have some meaningful callback to see the improvements as well.
Also share perf number compared to v8/sm.

---
In reply to: [411379017](https://github.com/Microsoft/ChakraCore/pull/5584#issuecomment-411379017) [](ancestors = 411379017)

**Comment by rhuanjl** [comment]
Here's a little benchmark that shows some slightly more realistic but nonetheless good results - still a little contrived but should at least ensure that the callback isn't optimised out:

```js
const glob = [];
const bar = [];

for (let i = 0; i < 1000; ++i)
{
    bar [i] = i * i;
}

function doSomething(a)
{
    glob.push(a * Math.random());
}

function test()
{
    const start = Date.now();
    for (let i = 0; i < 100000; ++i)
    {
        bar.forEach(doSomething);
        glob.length = 0;
    }
    return Date.now() - start;
}

let results = 0;

for (let j = 0; j < 20; ++j)
{
    results += test();
}

print (`Average time was ${results / 20} ms`);
```

eshost results (note chDev = CC with this change whereas Chakra = CC 1.10.1):
```
┌────────────────┬─────────────────────────────┐
│ JavaScriptCore │ Average time was 798.5 ms   │
├────────────────┼─────────────────────────────┤
│ SpiderMonkey   │ Average time was 905.1 ms   │
├────────────────┼─────────────────────────────┤
│ CC this branch │ Average time was 1034.25 ms │
├────────────────┼─────────────────────────────┤
│ V8 --harmony   │ Average time was 2243.95 ms │
├────────────────┼─────────────────────────────┤
│ CC 1.10.1      │ Average time was 2558.3 ms  │
└────────────────┴─────────────────────────────┘
```

**Comment by rhuanjl** [comment]
Having added a positive speed test I think I also need to add a negative one, add a second argument (i.e. a thisArg) to the forEach call in my benchmark above and the results change somewhat - CC with this change becomes the slowest by a distance (significantly slower than CC without this change).


| Engine        | Average time (ms)           |
| ------------- |:-------------:| 
| JavaScriptCore     | 751 | 
| SpiderMonkey      | 854.1      | 
| v8 | 2227.2      |  
| CC 1.10.1 | 2614.8      |  
| CC this branch | 4206.1      |  

CC's handling of bound functions is very slow - is there a reason this can't use callbackfn.call or a "safe" version of that? (aware that the Array.prototype.filter implementation uses bind as well it also has performance problems for anyone substituting their own thisArg.)

**Comment by MSLaguana** [comment]
We should be able to use the same pattern, `__chakraLibrary.callInstanceFunc(__chakraLibrary.functionApply, callbackfn, thisArg, argsList);` or something similar to get the equivalent of `callbackfn.apply(thisArg, argsList)` but avoiding issues with malformed functions / messed up prototypes etc. I don't know if that's going to necessarily be faster though, my understanding is that we don't do great with indirect function calls. If it was literally "callbackfn.apply(...)` I think we have some special casing to directly call the function (or inline it), but because we have to jump through hoops here 'just in case' I think the JIT doesn't have full information.

**Comment by sigatrev** [comment]
The reason we are worse with bound functions is that we cannot currently inline them. Without inlining, the JIT cannot be sure what those other functions are doing and so it has to repeat type checks and bound checks on every iteration of the loops. When the callbacks are inlined we're able to verify they don't modify the original array and we can usually hoist type and bound checks out of the loop.

We're able to inline function.call and function.apply but because of how those functions work, being part of the function prototype, we can't call them from the built ins without using additional indirection which doesn't get inlined. We do plan to look into inlining bound functions and/or some of the __chakraLibrary helper functions in the future.

As for the perf on cases where thisArg is provided, we believe that the usage of thisArg with forEach is highly uncommon, but if that assumption is incorrect we can look at inlining in the above mentioned conditions sooner rather than later.

**Comment by yullin-ms** [comment]
@dotnet-bot test Ubuntu shared_ubuntu_linux_debug now!

**Comment by jackhorton** [comment]
With that steep of a perf cliff for thisArg, couldnt we just fallback to JavascriptArray::EntryForEach if thisArg is provided? That would at least eliminate the regression and keep us ~on par with V8 in that scenario.

**Comment by sigatrev** [comment]
@jackhorton I'm planning to get the __chakraLibrary version of function.call to inline before this ships, but if I don't have the time then we can take that route as a backup

File: lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js (hunk 1 above)
Comment (yullin-ms) [reply lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js]: The code have been refactored

**Pull request merged**